### ISSUE TITLE
maybe exist a bug in mlt_events.c

### ISSUE THREAD

**Issue by davidwongiiss1984** [mltframework/mlt#443]
I need get event source when i write an event listener, but i find following code in mlt_events_fire method:
`transmitter( event->listener, event->owner, event->service, args );`
**event->owner is mlt_events object, not a mlt_service.**

I modified and works ok :
- change event->owner to **event->owner->owner**
- set events->owner = self in mlt_events_store, because it always NULL 
```
static void mlt_events_store(mlt_properties self, mlt_events events)
{
	if (self != NULL && events != NULL)
	{
		if (events->owner == NULL) {
			events->owner = self;
		}
		mlt_properties_set_data(self, "_events", events, 0, (mlt_destructor)mlt_events_close, NULL);
	}
}
```

**Comment by sway913** [comment]
Yes, I also encountered this problem

**Comment by ddennedy** [comment]
I kind of figured there is a reason literally nothing in MLT or Shotcuts uses this parameter. I was thinking it should be removed. Nearly all uses of `mlt_events_listen()` and the supplied `mlt_listener` supply something to the `void *service` parameter, which is received as the callback function's second argument.
It is much better to communicate by way of diff, patch, or pull request. Also, any significant change in the framework should not include a unit test in src/tests/ including the addition of a test suite if missing as in this case.
Honestly, there is not really any need for `mlt_events_store()` as a separate static function.
Here is my diff
```
--- src/framework/mlt_events.c
+++ src/framework/mlt_events.c
@@ -127,9 +127,8 @@ void mlt_event_close( mlt_event self )
 /* Forward declaration to private functions.
 */
 
 static mlt_events mlt_events_fetch( mlt_properties );
-static void mlt_events_store( mlt_properties, mlt_events );
 static void mlt_events_close( mlt_events );
 
 /** Initialise the events structure.
  *
@@ -139,13 +138,15 @@ static void mlt_events_close( mlt_events );
 
 void mlt_events_init( mlt_properties self )
 {
 	mlt_events events = mlt_events_fetch( self );
-	if ( events == NULL )
-	{
+	if (events && self) {
 		events = calloc( 1, sizeof( struct mlt_events_struct ) );
-		events->list = mlt_properties_new( );
-		mlt_events_store( self, events );
+		if (events) {
+			events->list = mlt_properties_new( );
+			events->owner = self;
+			mlt_properties_set_data( self, "_events", events, 0, ( mlt_destructor )mlt_events_close, NULL );
+		}
 	}
 }
 
 /** Register an event and transmitter.
@@ -212,11 +213,11 @@ int mlt_events_fire( mlt_properties self, const char *id, ... )
 				mlt_event event = mlt_properties_get_data_at( listeners, i, NULL );
 				if ( event != NULL && event->owner != NULL && event->block_count == 0 )
 				{
 					if ( transmitter != NULL )
-						transmitter( event->listener, event->owner, event->service, args );
+						transmitter( event->listener, event->owner->owner, event->service, args );
 					else
-						event->listener( event->owner, event->service );
+						event->listener( event->owner->owner, event->service );
 					++result;
 				}
 			}
 		}
@@ -471,21 +472,8 @@ static mlt_events mlt_events_fetch( mlt_properties self )
 		events = mlt_properties_get_data( self, "_events", NULL );
 	return events;
 }
 
-/** Store the events object.
- *
- * \private \memberof mlt_events_struct
- * \param self a properties list
- * \param events an events object
- */
-
-static void mlt_events_store( mlt_properties self, mlt_events events )
-{
-	if ( self != NULL && events != NULL )
-		mlt_properties_set_data( self, "_events", events, 0, ( mlt_destructor )mlt_events_close, NULL );
-}
-
 /** Close the events object.
  *
  * \private \memberof mlt_events_struct
  * \param events an events object
```

**Comment by ddennedy** [comment]
I think this change is safe since nothing in MLT is using the owner argument and mlt_events_struct has been private ever since it was created.

**Comment by davidwongiiss1984** [comment]
👍

**Comment by ddennedy** [comment]
Fixed in 4a94afd70ea278141530a885d08394f5a4090715. I got the bug number wrong on the commit message. :-(
